File: decibel-meter/android/app/src/main/java/com/izowooi/dm/core/Acoustics.kt
package com.izowooi.dm.core

import kotlin.math.*

enum class Weighting { A, FLAT }

object Acoustics {
    const val FLOOR_DB = -120.0
    const val WINDOW_MS = 100
    const val MAX_SECONDS = 4 * 60 * 60

    fun decibels(meanSquare: Double): Double =
        if (!meanSquare.isFinite() || meanSquare <= 1e-12) FLOOR_DB
        else max(FLOOR_DB, 10.0 * log10(meanSquare))

    fun rmsDb(samples: DoubleArray): Double {
        require(samples.isNotEmpty() && samples.all { it.isFinite() })
        return decibels(samples.sumOf { it * it } / samples.size)
    }

    fun energyAverage(levels: List<Pair<Double, Long>>): Double {
        require(levels.isNotEmpty() && levels.all { it.first.isFinite() && it.second > 0 })
        return decibels(levels.sumOf { 10.0.pow(it.first / 10.0) * it.second } / levels.sumOf { it.second })
    }
}

/** Stateful second-order sections, derived from the analogue A-weighting poles.
 * The digital response is normalized at 1 kHz. No microphone sensitivity is assumed.
 */
class FrequencyFilter(sampleRate: Double, weighting: Weighting) {
    private data class Section(
        val b0: Double, val b1: Double, val b2: Double,
        val a1: Double, val a2: Double,
        var z1: Double = 0.0, var z2: Double = 0.0,
    ) {
        fun apply(x: Double): Double {
            val y = b0 * x + z1
            z1 = b1 * x - a1 * y + z2
            z2 = b2 * x - a2 * y
            return y
        }
        fun magnitude(omega: Double): Double {
            val nReal = b0 + b1 * cos(omega) + b2 * cos(2 * omega)
            val nImag = -b1 * sin(omega) - b2 * sin(2 * omega)
            val dReal = 1 + a1 * cos(omega) + a2 * cos(2 * omega)
            val dImag = -a1 * sin(omega) - a2 * sin(2 * omega)
            return sqrt((nReal * nReal + nImag * nImag) / (dReal * dReal + dImag * dImag))
        }
    }

    private val sections: List<Section>
    private val gain: Double

    init {
        require(sampleRate >= 32000 && sampleRate <= 192000)
        sections = if (weighting == Weighting.FLAT) emptyList() else {
            fun pole(hz: Double): Double = (2 * sampleRate - 2 * PI * hz) / (2 * sampleRate + 2 * PI * hz)
            fun section(zero: Double, p1: Double, p2: Double) = Section(1.0, -2 * zero, 1.0, -(p1 + p2), p1 * p2)
            listOf(
                section(1.0, pole(20.598997), pole(20.598997)),
                section(1.0, pole(107.65265), pole(737.86223)),
                section(-1.0, pole(12194.217), pole(12194.217)),
            )
        }
        gain = 1.0 / sections.fold(1.0) { value, section -> value * section.magnitude(2 * PI * 1000 / sampleRate) }
    }

    fun process(sample: Double): Double {
        var value = sample * gain
        sections.forEach { value = it.apply(value) }
        return value
    }
}

data class AudioWindow(
    val endSample: Long,
    val count: Int,
    val squareSum: Double,
    val peak: Double,
    val clipped: Int,
    val invalid: Int,
) {
    val dbfs: Double get() = Acoustics.decibels(squareSum / count)
}

/** Single audio-thread owner. Emits only summaries; PCM never leaves the audio engine. */
class WindowProcessor(val sampleRate: Int, weighting: Weighting) {
    private val filter = FrequencyFilter(sampleRate.toDouble(), weighting)
    private val windowSize = sampleRate / 10
    private var count = 0
    private var squareSum = 0.0
    private var peak = 0.0
    private var clipped = 0
    private var invalid = 0
    private var totalSamples = 0L

    fun accept(sample: Double): AudioWindow? {
        val safe = if (sample.isFinite()) sample else { invalid++; 0.0 }
        if (abs(safe) >= 0.999) clipped++
        peak = max(peak, abs(safe))
        val weighted = filter.process(safe)
        squareSum += weighted * weighted
        count++
        totalSamples++
        return if (count == windowSize) flush() else null
    }

    fun flush(): AudioWindow? {
        if (count == 0) return null
        val result = AudioWindow(totalSamples, count, squareSum, peak, clipped, invalid)
        count = 0; squareSum = 0.0; peak = 0.0; clipped = 0; invalid = 0
        return result
    }
}

data class InputConditions(
    val device: String,
    val routeId: String,
    val routeName: String,
    val sampleRate: Int,
    val channels: Int,
    val source: String,
    val weighting: Weighting,
    val gain: String = "system-fixed",
    val pipeline: String = "sori-v1:pcm-normalized:100ms:a-bilinear-1khz",
) {
    fun matches(other: InputConditions): Boolean = this == other
}

data class CalibrationProfile(
    val id: String,
    val name: String,
    val createdAt: String,
    val referenceDb: Double,
    val measuredDbfs: Double,
    val notes: String,
    val conditions: InputConditions,
) {
    val offset: Double get() = referenceDb - measuredDbfs
    fun validFor(input: InputConditions): Boolean =
        conditions.matches(input) && referenceDb.isFinite() && referenceDb in 20.0..140.0 &&
            measuredDbfs.isFinite() && measuredDbfs > -100.0 && measuredDbfs < 0.0 && offset in 0.0..180.0
}

data class LevelPoint(val seconds: Double, val dbfs: Double, val count: Int, val clipped: Int = 0)

data class MeterSnapshot(
    val current: Double = Acoustics.FLOOR_DB,
    val minimum: Double = Acoustics.FLOOR_DB,
    val average: Double = Acoustics.FLOOR_DB,
    val maximum: Double = Acoustics.FLOOR_DB,
    val seconds: Double = 0.0,
    val count: Long = 0,
    val clipped: Long = 0,
    val currentClipped: Boolean = false,
    val points: List<LevelPoint> = emptyList(),
) {
    val canCalibrate: Boolean get() {
        val recent = points.takeLast(30)
        if (recent.size < 30 || recent.any { it.clipped > 0 || it.dbfs <= -100 }) return false
        val mean = recent.sumOf { it.dbfs } / recent.size
        return sqrt(recent.sumOf { (it.dbfs - mean).pow(2) } / recent.size) < 2.0
    }
    val calibrationLevel: Double get() = Acoustics.energyAverage(points.takeLast(30).map { it.dbfs to it.count.toLong() })
}

class SessionAccumulator(private val sampleRate: Int) {
    private val graph = ArrayDeque<LevelPoint>()
    private val export = ArrayList<LevelPoint>()
    private var count = 0L
    private var energy = 0.0
    private var minimum = Double.POSITIVE_INFINITY
    private var maximum = Double.NEGATIVE_INFINITY
    private var clipped = 0L
    private var secondEnergy = 0.0
    private var secondCount = 0
    private var secondClipped = 0
    var snapshot = MeterSnapshot()
        private set

    fun add(window: AudioWindow): MeterSnapshot {
        require(window.count > 0 && window.invalid == 0 && window.squareSum.isFinite())
        require(count + window.count <= sampleRate.toLong() * Acoustics.MAX_SECONDS)
        count += window.count
        energy += window.squareSum
        clipped += window.clipped
        minimum = min(minimum, window.dbfs)
        maximum = max(maximum, window.dbfs)
        val seconds = count.toDouble() / sampleRate
        graph.addLast(LevelPoint(seconds, window.dbfs, window.count, window.clipped))
        while (graph.size > 600) graph.removeFirst()
        secondEnergy += window.squareSum; secondCount += window.count; secondClipped += window.clipped
        if (secondCount >= sampleRate) flushExport(seconds)
        snapshot = MeterSnapshot(window.dbfs, minimum, Acoustics.decibels(energy / count), maximum,
            seconds, count, clipped, window.clipped > 0, graph.toList())
        return snapshot
    }

    private fun flushExport(seconds: Double) {
        if (secondCount == 0) return
        export.add(LevelPoint(seconds, Acoustics.decibels(secondEnergy / secondCount), secondCount, secondClipped))
        secondEnergy = 0.0; secondCount = 0; secondClipped = 0
    }

    fun finish(): List<LevelPoint> {
        flushExport(count.toDouble() / sampleRate)
        return export.toList()
    }
}

File: decibel-meter/ios/dm/Acoustics.swift
import Foundation

enum Weighting: String, Codable, CaseIterable, Sendable { case a = "A", flat = "FLAT" }

enum Acoustics {
    static let floorDB = -120.0
    static let windowMS = 100
    static let maximumSeconds = 4 * 60 * 60

    static func decibels(_ meanSquare: Double) -> Double {
        guard meanSquare.isFinite, meanSquare > 1e-12 else { return floorDB }
        return max(floorDB, 10 * log10(meanSquare))
    }

    static func rmsDB(_ samples: [Double]) -> Double? {
        guard !samples.isEmpty, samples.allSatisfy(\.isFinite) else { return nil }
        return decibels(samples.reduce(0) { $0 + $1 * $1 } / Double(samples.count))
    }

    static func energyAverage(_ levels: [(Double, Int)]) -> Double? {
        guard !levels.isEmpty, levels.allSatisfy({ $0.0.isFinite && $0.1 > 0 }) else { return nil }
        let count = levels.reduce(0) { $0 + $1.1 }
        return decibels(levels.reduce(0) { $0 + pow(10, $1.0 / 10) * Double($1.1) } / Double(count))
    }
}

/// Analogue A-weighting poles transformed into digital second-order sections.
/// Normalizing at 1 kHz is a filter gain, never a microphone calibration offset.
struct FrequencyFilter {
    private struct Section {
        let b0: Double, b1: Double, b2: Double, a1: Double, a2: Double
        var z1 = 0.0, z2 = 0.0

        mutating func process(_ x: Double) -> Double {
            let y = b0 * x + z1
            z1 = b1 * x - a1 * y + z2
            z2 = b2 * x - a2 * y
            return y
        }

        func magnitude(_ omega: Double) -> Double {
            let nr = b0 + b1 * cos(omega) + b2 * cos(2 * omega)
            let ni = -b1 * sin(omega) - b2 * sin(2 * omega)
            let dr = 1 + a1 * cos(omega) + a2 * cos(2 * omega)
            let di = -a1 * sin(omega) - a2 * sin(2 * omega)
            return sqrt((nr * nr + ni * ni) / (dr * dr + di * di))
        }
    }

    private var sections: [Section]
    private let gain: Double

    init(sampleRate: Double, weighting: Weighting) {
        precondition((32000...192000).contains(sampleRate))
        if weighting == .flat {
            sections = []
        } else {
            func pole(_ hz: Double) -> Double { (2 * sampleRate - 2 * .pi * hz) / (2 * sampleRate + 2 * .pi * hz) }
            func section(_ zero: Double, _ p1: Double, _ p2: Double) -> Section {
                Section(b0: 1, b1: -2 * zero, b2: 1, a1: -(p1 + p2), a2: p1 * p2)
            }
            sections = [
                section(1, pole(20.598997), pole(20.598997)),
                section(1, pole(107.65265), pole(737.86223)),
                section(-1, pole(12194.217), pole(12194.217))
            ]
        }
        gain = 1 / sections.reduce(1) { $0 * $1.magnitude(2 * .pi * 1000 / sampleRate) }
    }

    mutating func process(_ sample: Double) -> Double {
        var value = sample * gain
        for index in sections.indices { value = sections[index].process(value) }
        return value
    }
}

struct AudioWindow: Sendable {
    let endSample: Int64
    let count: Int
    let squareSum: Double
    let peak: Double
    let clipped: Int
    let invalid: Int
    var dbfs: Double { Acoustics.decibels(squareSum / Double(count)) }
}

/// Owned by the serial audio callback; raw PCM is never stored or dispatched to UI.
final class WindowProcessor {
    let sampleRate: Int
    private var filter: FrequencyFilter
    private let windowSize: Int
    private var count = 0, clipped = 0, invalid = 0
    private var squareSum = 0.0, peak = 0.0
    private var totalSamples: Int64 = 0

    init(sampleRate: Int, weighting: Weighting) {
        self.sampleRate = sampleRate
        filter = FrequencyFilter(sampleRate: Double(sampleRate), weighting: weighting)
        windowSize = sampleRate / 10
    }

    func accept(_ sample: Double) -> AudioWindow? {
        let safe: Double
        if sample.isFinite { safe = sample } else { invalid += 1; safe = 0 }
        if abs(safe) >= 0.999 { clipped += 1 }
        peak = max(peak, abs(safe))
        let weighted = filter.process(safe)
        squareSum += weighted * weighted
        count += 1
        totalSamples += 1
        return count == windowSize ? flush() : nil
    }

    func flush() -> AudioWindow? {
        guard count > 0 else { return nil }
        let result = AudioWindow(endSample: totalSamples, count: count, squareSum: squareSum,
                                 peak: peak, clipped: clipped, invalid: invalid)
        count = 0; squareSum = 0; peak = 0; clipped = 0; invalid = 0
        return result
    }
}

struct InputConditions: Codable, Equatable, Sendable {
    let device: String
    let routeID: String
    let routeName: String
    let sampleRate: Int
    let channels: Int
    let source: String
    let weighting: Weighting
    let gain: String
    var pipeline = "sori-v1:pcm-normalized:100ms:a-bilinear-1khz"
}

struct CalibrationProfile: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let createdAt: String
    let referenceDB: Double
    let measuredDBFS: Double
    let notes: String
    let conditions: InputConditions
    var offset: Double { referenceDB - measuredDBFS }

    func valid(for input: InputConditions) -> Bool {
        conditions == input && referenceDB.isFinite && (20...140).contains(referenceDB) &&
        measuredDBFS.isFinite && measuredDBFS > -100 && measuredDBFS < 0 && (0...180).contains(offset)
    }
}

struct LevelPoint: Codable, Sendable {
    let seconds: Double
    let dbfs: Double
    let count: Int
    var clipped: Int = 0
}

struct MeterSnapshot: Codable, Sendable {
    var current = Acoustics.floorDB
    var minimum = Acoustics.floorDB
    var average = Acoustics.floorDB
    var maximum = Acoustics.floorDB
    var seconds = 0.0
    var count: Int64 = 0
    var clipped: Int64 = 0
    var currentClipped = false
    var points: [LevelPoint] = []

    var canCalibrate: Bool {
        let recent = points.suffix(30)
        guard recent.count == 30, recent.allSatisfy({ $0.clipped == 0 && $0.dbfs > -100 }) else { return false }
        let mean = recent.reduce(0) { $0 + $1.dbfs } / Double(recent.count)
        return sqrt(recent.reduce(0) { $0 + pow($1.dbfs - mean, 2) } / Double(recent.count)) < 2
    }

    var calibrationLevel: Double? {
        Acoustics.energyAverage(points.suffix(30).map { ($0.dbfs, $0.count) })
    }
}

struct SessionAccumulator {
    let sampleRate: Int
    init(sampleRate: Int) { self.sampleRate = sampleRate }
    private var graph: [LevelPoint] = []
    private var export: [LevelPoint] = []
    private var count: Int64 = 0
    private var energy = 0.0
    private var minimum = Double.infinity, maximum = -Double.infinity
    private var clipped: Int64 = 0
    private var secondEnergy = 0.0
    private var secondCount = 0, secondClipped = 0
    private(set) var snapshot = MeterSnapshot()

    mutating func add(_ window: AudioWindow) -> MeterSnapshot? {
        guard window.count > 0, window.invalid == 0, window.squareSum.isFinite,
              count + Int64(window.count) <= Int64(sampleRate * Acoustics.maximumSeconds) else { return nil }
        count += Int64(window.count)
        energy += window.squareSum
        clipped += Int64(window.clipped)
        minimum = min(minimum, window.dbfs); maximum = max(maximum, window.dbfs)
        let seconds = Double(count) / Double(sampleRate)
        graph.append(LevelPoint(seconds: seconds, dbfs: window.dbfs, count: window.count, clipped: window.clipped))
        if graph.count > 600 { graph.removeFirst() }
        secondEnergy += window.squareSum; secondCount += window.count; secondClipped += window.clipped
        if secondCount >= sampleRate { flushExport(seconds) }
        snapshot = MeterSnapshot(current: window.dbfs, minimum: minimum,
            average: Acoustics.decibels(energy / Double(count)), maximum: maximum,
            seconds: seconds, count: count, clipped: clipped, currentClipped: window.clipped > 0, points: graph)
        return snapshot
    }

    private mutating func flushExport(_ seconds: Double) {
        guard secondCount > 0 else { return }
        export.append(LevelPoint(seconds: seconds, dbfs: Acoustics.decibels(secondEnergy / Double(secondCount)),
                                 count: secondCount, clipped: secondClipped))
        secondEnergy = 0; secondCount = 0; secondClipped = 0
    }

    mutating func finish() -> [LevelPoint] {
        flushExport(Double(count) / Double(sampleRate))
        return export
    }
}
